LiS/man/man8/thrtst.8.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) $Id$
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2009  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date$ by $Author$
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log: thrtst.8.man,v $
.\" Revision 0.9.2.4  2008-04-29 08:33:20  brian
.\" - update headers for Affero release
.\"
.\" Revision 0.9.2.3  2006/04/05 10:05:02  brian
.\" - get references in right place
.\"
.\" Revision 0.9.2.2  2006/04/05 09:42:41  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\"
.\"
.TH THRTST 8 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ System Admin."
.SH NAME
.B thrtst
\- thread tests for LiS STREAMS package
.\"
.\"
.SH SYNOPSIS
.PP
.\".HP 12
.\"\fBthrtst\fR [\fIcommand\fR]
.\".PD 0
.HP 12
\fBthrtst\fR [\fIoptions\fR]
.PD 0
.HP 12
\fBthrtst\fR {\fB-h\fR,\fB--help\fR}
.HP 12
\fBthrtst\fR {\fB-V\fR,\fB--version\fR}
.HP 12
\fBthrtst\fR {\fB--copying\fR}
.PD
.\"
.\"
.SH DESCRIPTION
.PP
.B thrtst
is a test program for testing the LiS STREAMS package.
.\"
.\"
.\".SH COMMANDS
.\".PP
.\"
.\"
.SH OPTIONS
.PP
.TP
\fB-v\fP, \fB--verbose\fP [\fILEVEL\fP]
increase verbosity or set to \fILEVEL\fP.  The default is one (1).
This option may be repeated.
.TP
\fB-q\fP, \fB--quiet\fP
Suppress normal output (equivalent to \fB--verbose\fP=\fI0\fP).
.TP
\fB-h\fP, \fB--help\fP, \fB-?\fP, \fB--?\fP
print a usage message and exit.
.TP
\fB-V\fP, \fB--version\fP
print the version and exit.
.TP
\fB-C\fP, \fB--copying\fP
print copying permission and exit.
.\"
.\"
.\".SH DIAGNOSTICS
.\".PP
.\"
.\"
.SH BUGS
.PP
.B thrtst
has no known bugs.
.\"
.\"
.\".SH EXAMPLES
.\".PP
.\"
.\"
.SH "SEE ALSO"
.PP
.BR oc (1),
.BR rwf (1),
.BR pafd (1),
.BR psfd (1),
.BR lis_test (8),
.BR lis_test2 (8),
.BR test-streams (8),
.BR strtst (8),
.BR timetst (8),
.BR STREAMS (4),
.IR "Linux STREAMS Installation and Reference Manual" .
.\"
.\"
.SH COMPATIBILITY
.PP
.BR thrtst
is compatible with
.B LiS\c
.
.[
LiS
.]
.\"
.\"
.SH CONFORMANCE
.PP
.BR thrtst
does not conform to any standard.
.B thrtst
tests conformance of
.BR STREAMS (4)
to the specifications listed on the
.BR STREAMS (4)
manual page.
.\"
.\"
.SH VERSIONS
.PP
This manpage was written for
.B LiS\c
.
.[
LiS
.]
.\"
.\"
.SH HISTORY
.PP
.B thrtst
is provided with the
.I @PACKAGE_TITLE@
package.
.[
$LIST$
.]
.TI
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77


NEWS
OpenSS7 Master Package -- history of user visible changes.  2009-05-04
$Id: NEWS,v 0.9.2.72 2009-04-19 00:28:25 brian Exp $
Copyright (c) 2008-2009  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
`OpenSS7 Master Package' package.

 Major changes for release openss7-0.9.2.F.20090405
---------------------------------------------------

This is the seventh `OpenSS7 Project' release of the `OpenSS7 Master
Package' package.  For a sub-package listing, see `README'
(http://www.openss7.org/codefiles/openss7/README) in the release.  Most
of the sub-packages in the release are production grade for `Linux
Fast-STREAMS'.(1)  All existing validation test suites run clean on
supported distributions, production kernels and architectures.

The original intention was that the `OpenSS7 Master Package' package
not be released as often as subpackages and, indeed, newer subpackages
can be strapped into the master package and built automatically.
However, in reality, the `OpenSS7 Master Package' package has been
released as often as the subpackages.  To rebuild the master package
with a new sub-package release, simply replace the directory to which
the sub-package belongs with the unpackged sub-package reelase and then
rebuild the master package.  This release provides support for recent
distributions, production kernels and tool chains.  The release also
provides support for significiant new features in the sub-packages.

Major features since the last public release are:

   - Deprecated sub-packages have been excluded from the master package
     release.  The excluded sub-packages are:

         - `sctp-0.2.27.20090405'
         - `iperf-2.0.8.20090405'
         - `netperf-2.3.7.20090405'
         - `osr61-0.9.2.3.20090405'

   - Documentation license upgraded to GFDL 1.3.

   - Updated tool chain to `m4-1.4.13', `automake-1.10.2',
     `bison-2.4.1', `libtool-2.2.6a', `texinfo-4.13a'.

   - Addition of the following sub-packages:

         - `stratm-0.9.2.1.20090405'

   - Addition of the `stratm' sub-package to the master build and
     release that now includes all of the ATM components that were
     previously part of the `strss7' sub-package.

   - Significant feature updates were made to:

         - `strchan-0.9.2.4.20090405'
         - `strsock-0.9.2.4.20090405'
         - `strnsl-0.9.2.4.20090405'
         - `sigtran-0.9.2.4.20090405'

   - Significant rework of the `strchan' sub-package into the `strss7',
     `strisdn', `strx25' and `striso' packages.  One single driver now
     meets the needs of all sub-packages.

   - Completion of the `strsock' package, moves from pre-alpha to beta.

   - The `strnsl' subpackage moves from alpha to production.

   - Significant completion of the `sigtran' package: all components
     moved back to production.


_This is a public, stable, production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
`rpm(8)' and `dpkg(1)' based distributions, and can be used on
production kernels without patching or recompiling the kernel.

This package is publicly released strictly under the `GNU Affero
General Public License Version 3'; however, commercial licensing is
available.  The release is available as an `autoconf(1)' tarball, SRPM,
DSC, and a set of binary RPMs and DEBs for popular distributions.  See
the package page (http://www.openss.org/openss7_pkg.html) for the
`autoconf(1)' tarballs, SRPMs, DSCs and binary RPMs and DEBs.  See the
repository page (http://www.openss7.org/repos.html) for information
concerning network installation and update sources.

See
`http://www.openss7.org/codefiles/openss7-0.9.2.F.20090405/ChangeLog'
and `http://www.openss7.org/codefiles/openss7-0.9.2.F.20090405/NEWS' in
the release for more information.  Also, see the `openss7.pdf' manual
in the release (also in html
(http://www.openss7.org/openss7_manual.html)).

For the news release, see `http://www.openss7.org/rel2009XXXX_L.html'.

---------- Footnotes ----------

(1) Sub-packages are no longer supported on the deprecated `LiS' STREAMS
package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - License upgrade to `AGPL Version 3'.


LiS/util/linux/thrtst.c
/*****************************************************************************

 @(#) $RCSfile: thrtst.c,v $ $Name:  $($Revision: 1.1.1.1.12.8 $) $Date: 2008-04-29 08:33:28 $

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>, or write to the
 Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date: 2008-04-29 08:33:28 $ by $Author: brian $

 -----------------------------------------------------------------------------

 $Log: thrtst.c,v $
 Revision 1.1.1.1.12.8  2008-04-29 08:33:28  brian
 - update headers for Affero release

 Revision 1.1.1.1.12.7  2007/08/14 10:47:33  brian
 - GPLv3 header update

 Revision 1.1.1.1.12.3  2005/05/14 08:35:16  brian
 - copyright header correction

 *****************************************************************************/

#ident "@(#) $RCSfile: thrtst.c,v $ $Name:  $($Revision: 1.1.1.1.12.8 $) $Date: 2008-04-29 08:33:28 $"

static char const ident[] =
    "$RCSfile: thrtst.c,v $ $Name:  $($Revision: 1.1.1.1.12.8 $) $Date: 2008-04-29 08:33:28 $";

/************************************************************************
*                     Multi-Threaded Test                               *
*************************************************************************
*									*
* This program runs data through the STREAMS loopback driver using	*
* multiple threads.  The threads act in pairs with one thread of each	*
* pair generating data and the other thread reading data.  The "reading"*
* thread echos the data back so that the generating thread can read	*
* back what it originally wrote.					*
*									*
* The main program monitors the progress and maintains a summary screen.*
*									*
************************************************************************/

/* 
 * Copyright (C)  2001  David Grothe, Gcom, Inc <dave@gcom.com>
 */

#define _REENTRANT
#define _THREAD_SAFE
#define _XOPEN_SOURCE	500	/* single unix spec */

#ifdef _GNU_SOURCE
#include <getopt.h>
#endif				/* _GNU_SOURCE */
#include <stropts.h>
#include <stddef.h>
#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <pthread.h>
#include <sys/ioctl.h>
#ifdef _GNU_SOURCE
#include <getopt.h>
#endif				/* _GNU_SOURCE */
#include <sys/LiS/loop.h>	/* an odd place for this file */

int verbose = 1;

/************************************************************************
*                          Constants                                    *
************************************************************************/
#define MAX_THR		16	/* less than # of loop-clones */
#define	MAX_MSG		1024

typedef struct thread_info {
	int thread_number;
	pthread_mutex_t startup;	/* so everyone can start at once */

	int xmcnt;			/* transmit msg count */
	int xbcnt;			/* transmit byte count */
	int rmcnt;			/* receive msg count */
	int rbcnt;			/* receive byte count */

	int fd;				/* file descr to loop driver */
	int minor;			/* minor device number of loop */
	int looped;			/* have been looped to partner */
	int burst;			/* initial burst size for writer */

	char *xdbuf;			/* xmit data buffer */
	char *xcbuf;			/* xmit ctl buffer */
	char *rdbuf;			/* receive data buffer */
	char *rcbuf;			/* receive ctl buffer */
	struct strbuf xd;
	struct strbuf xc;
	struct strbuf rd;
	struct strbuf rc;

	struct thread_info *other;	/* pointer to partner's struct */

} thread_info_t;

/************************************************************************
*                       Global Storage                                  *
************************************************************************/
pthread_t thread_ids[MAX_THR + 1];	/* numbered from 1 */
thread_info_t thread_info[MAX_THR + 1];	/* numbered from 1 */
time_t start_time;
int max_threads = MAX_THR;
int time_factor = 1;
int initial_burst = 0;

/************************************************************************
*                           setup_thread_info				*
*************************************************************************
*									*
* Initialize a thread info structure.					*
*									*
************************************************************************/
void
setup_thread_info(int thread_nr)
{
	thread_info_t *tp = &thread_info[thread_nr];
	int i;
	char *p;
	int arg;
	int rslt;
	int other_inx;			/* thr nr of my partner */
	struct strioctl ioc;
	pthread_mutexattr_t attr;

	memset(tp, 0, sizeof(*tp));
	tp->thread_number = thread_nr;
	tp->burst = initial_burst;

	tp->fd = open("/dev/loop_clone", O_RDWR, 0);
	if (tp->fd < 0) {
		if (verbose)
			perror("/dev/loop_clone");
		exit(1);
	}

	ioc.ic_timout = 10;
	ioc.ic_len = sizeof(int);
	ioc.ic_dp = (char *) &arg;

	ioc.ic_cmd = LOOP_GET_DEV;
	arg = -1;
	rslt = ioctl(tp->fd, I_STR, &ioc);
	if (rslt < 0) {
		if (verbose)
			perror("loop_clone: ioctl LOOP_GET_DEV");
		exit(1);
	}

	if (arg < 0) {
		if (verbose)
			fprintf(stderr, "loop_clone: ioctl LOOP_GET_DEV returned %d\n", arg);
		exit(1);
	}

	tp->minor = arg;	/* minor number of this loop clone */
	other_inx = ((tp->thread_number - 1) ^ 1) + 1;
	tp->other = &thread_info[other_inx];

	tp->xdbuf = malloc(MAX_MSG);
	tp->xcbuf = malloc(MAX_MSG);
	tp->rdbuf = malloc(MAX_MSG);
	tp->rcbuf = malloc(MAX_MSG);
	if (tp->xdbuf == NULL || tp->xcbuf == NULL || tp->rdbuf == NULL || tp->rcbuf == NULL) {
		if (verbose)
			fprintf(stderr, "setup_thread_info: cannot allocate memory\n");
		exit(1);
	}

	tp->xd.buf = tp->xdbuf;
	tp->xc.buf = tp->xcbuf;
	tp->rd.buf = tp->rdbuf;
	tp->rc.buf = tp->rcbuf;
	tp->rd.maxlen = MAX_MSG;
	tp->rc.maxlen = MAX_MSG;

	/* 
	 * Place a counting pattern in the xmit buffers
	 */
	for (p = tp->xdbuf, i = 0; i < MAX_MSG; i++, p++)
		*p = (char) i;
	for (p = tp->xcbuf, i = 0; i < MAX_MSG; i++, p++)
		*p = (char) i;

	pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_DEFAULT);
	pthread_mutex_init(&tp->startup, &attr);
	pthread_mutex_lock(&tp->startup);

}

/************************************************************************
*                         set_loop                                      *
*************************************************************************
*									*
* Set the loopback file to point to its partner.			*
*									*
************************************************************************/
int
set_loop(int fd, int other_minor)
{
	int rslt;
	struct strioctl ioc;

	if (verbose > 1)
		printf("set_loop(%d, %d)\n", fd, other_minor);
	ioc.ic_timout = 10;
	ioc.ic_cmd = LOOP_SET;
	ioc.ic_len = sizeof(int);
	ioc.ic_dp = (char *) &other_minor;
	rslt = ioctl(fd, I_STR, &ioc);
	if (rslt < 0)
		if (verbose)
			perror("set_loop: ioctl LOOP_SET");

	return (rslt);
}

/************************************************************************
*                           set_timer                                   *
*************************************************************************
*									*
* Set the timer for delay of messages in loopback driver.  Once set	*
* this value is used until the file is closed.				*
*									*
************************************************************************/
int
set_timer(int fd, int timeout)
{
	int rslt;
	struct strioctl ioc;

	timeout = time_factor * timeout;
	ioc.ic_cmd = LOOP_TIMR;	/* set timer for queue */
	ioc.ic_len = sizeof(int);
	ioc.ic_dp = (char *) &timeout;
	rslt = ioctl(fd, I_STR, &ioc);
	if (rslt < 0)
		if (verbose)
			perror("set_timer: ioctl LOOP_TIMR");

	return (rslt);
}

/************************************************************************
*                           reader                                      *
*************************************************************************
*									*
* The reader process.  It reads messages and writes them back.		*
*									*
************************************************************************/
void *
reader(void *arg)
{
	thread_info_t *tp = (thread_info_t *) arg;
	int nbytes;

	if (verbose > 1)
		printf("Reader[%d]: Waiting for startup\n", tp->thread_number);
	pthread_mutex_lock(&tp->startup);
	if (verbose)
		printf("Reader[%d]: Startup\n", tp->thread_number);

	for (;;) {
		if ((nbytes = read(tp->fd, tp->rdbuf, MAX_MSG)) < 0)
			break;

		tp->rmcnt++;
		tp->rbcnt += nbytes;

		if (write(tp->fd, tp->xdbuf, nbytes) < 0)
			break;

		tp->xmcnt++;
		tp->xbcnt += nbytes;
	}

	if (verbose)
		printf("Reader[%d]: Exiting\n", tp->thread_number);
	return (NULL);
}

/************************************************************************
*                           writer                                      *
*************************************************************************
*									*
* The writer process.  It writes messages and then reads them back.	*
*									*
************************************************************************/
void *
writer(void *arg)
{
	thread_info_t *tp = (thread_info_t *) arg;
	int msg_size = MAX_MSG;
	int i;
	int nbytes;

	if (verbose > 1)
		printf("Writer[%d]: Waiting for startup\n", tp->thread_number);
	pthread_mutex_lock(&tp->startup);
	if (verbose)
		printf("Writer[%d]: Startup\n", tp->thread_number);

	for (i = 0; i < tp->burst; i++) {
		if (write(tp->fd, tp->xdbuf, msg_size) < 0)
			break;

		tp->xmcnt++;
		tp->xbcnt += msg_size;
	}

	for (;;) {
		if (write(tp->fd, tp->xdbuf, msg_size) < 0)
			break;

		tp->xmcnt++;
		tp->xbcnt += msg_size;

		if ((nbytes = read(tp->fd, tp->rdbuf, MAX_MSG)) < 0)
			break;

		tp->rmcnt++;
		tp->rbcnt += nbytes;
	}

	if (verbose)
		printf("Writer[%d]: Exiting\n", tp->thread_number);
	return (NULL);
}

/************************************************************************
*                            print_status                               *
*************************************************************************
*									*
* Print a status report on the progress of the threads.			*
*									*
************************************************************************/
void
print_progress(void)
{
	int i;
	time_t now;
	long delta;
	thread_info_t *tp;

	if (verbose)
		printf("Thr   Xmit-Cnt Xmit/Sec    Rcv-Cnt  Rcv/Sec\n");
	now = time(NULL);
	delta = now - start_time;
	if (delta == 0)
		delta = 1;
	for (i = 1; i <= max_threads; i++) {
		tp = &thread_info[i];
		if (verbose)
			printf("%3u %10u %8u %10u %8u\n", i, tp->xmcnt, (int) (tp->xmcnt / delta),
			       tp->rmcnt, (int) (tp->rmcnt / delta));
	}
}

/************************************************************************
*                           get_options                                 *
************************************************************************/
void
copying(int argc, char *argv[])
{
	if (!verbose)
		return;
	fprintf(stdout, "\
\n\
%1$s %2$s:\n\
\n\
Copyright (c) 2003-2008  OpenSS7 Corporation <http://www.openss7.com/>\n\
Copyright (c) 2001       David Grothe, Gcom, Inc <dave@gcom.com>\n\
\n\
All Rights Reserved.\n\
\n\
This program is free software;  you can  redistribute  it and/or modify it under\n\
the terms of the GNU General  Public License as  published by the  Free Software\n\
Foundation; version 3 of the License.\n\
\n\
This program is distributed in the hope that it will be  useful, but WITHOUT ANY\n\
WARRANTY;  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A\n\
PARTICULAR PURPOSE.  See the GNU General Public License for more details.\n\
\n\
You should  have received  a copy of the GNU  General  Public License along with\n\
this program.  If not, see <http://www.gnu.org/licenses/>,  or write to the Free\n\
Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.\n\
\n\
U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on behalf\n\
of the  U.S. Government  (\"Government\"),  the following provisions apply to you.\n\
If the Software is  supplied by the Department of Defense (\"DoD\"), it is classi-\n\
fied as  \"Commercial Computer Software\"  under paragraph 252.227-7014 of the DoD\n\
Supplement  to the  Federal Acquisition Regulations  (\"DFARS\") (or any successor\n\
regulations) and the  Government  is acquiring  only the license rights  granted\n\
herein (the license  rights customarily  provided to non-Government  users).  If\n\
the Software is supplied to any unit or agency of the Government other than DoD,\n\
it is classified as  \"Restricted Computer Software\" and the  Government's rights\n\
in the  Software are defined in  paragraph 52.227-19 of the Federal  Acquisition\n\
Regulations  (\"FAR\") (or any successor regulations) or, in the cases of NASA, in\n\
paragraph  18.52.227-86 of the  NASA Supplement  to the  FAR (or  any  successor\n\
regulations).\n\
\n\
", argv[0], ident);
}

void
version(int argc, char *argv[])
{
	if (!verbose)
		return;
	fprintf(stdout, "\
\n\
%1$s %2$s:\n\
    Copyright (c) 2003-2008  OpenSS7 Corporation.  All Rights Reserved.\n\
    Copyright (c) 2001       David Grothe, Gcom, Inc <dave@gcom.com>\n\
\n\
    Distributed by OpenSS7 Corporation under GPL Version 3,\n\
    included here by reference.\n\
\n\
    See `%1$s --copying' for copying permissions.\n\
\n\
", argv[0], ident);
}

void
usage(int argc, char *argv[])
{
	if (!verbose)
		return;
	fprintf(stderr, "\
Usage:\n\
    %1$s [options]\n\
    %1$s {-h,--help}\n\
    %1$s {-V,--version}\n\
    %1$s {-C,--copying}\n\
", argv[0]);
}

void
help(int argc, char *argv[])
{
	if (!verbose)
		return;
	fprintf(stdout, "\
\n\
Usage:\n\
    %1$s [options]\n\
    %1$s {-h,--help}\n\
    %1$s {-V,--version}\n\
    %1$s {-C,--copying}\n\
Options:\n\
    -b, --burst BURST\n\
        set initial burst for writer thread\n\
    -f, --factor FACTOR\n\
        set time delay factor to FACTOR\n\
    -t, --threads THREADS\n\
        set number of threads to THREADS\n\
    -v, --verbose [LEVEL]\n\
        increase verbosity or set to LEVEL [default: 1]\n\
        this option may be repeated.\n\
    -q, --quiet\n\
        suppress normal output (equivalent to --verbose=0)\n\
    -h, --help, -?, --?\n\
        print this usage message and exit\n\
    -V, --version\n\
        print version and exit\n\
    -C, --copying\n\
        print copying permission and exit\n\
\n\
", argv[0]);
}

void
get_options(int argc, char **argv)
{
	for (;;) {
		int c;

#ifdef _GNU_SOURCE
		int option_index = 0;
		/* *INDENT-OFF* */
		static struct option long_options[] = {
			{ "burst",   required_argument,	NULL, 'b'},
			{ "factor",  required_argument,	NULL, 'f'},
			{ "threads", required_argument,	NULL, 't'},
			{ "quiet",   no_argument,	NULL, 'q'},
			{ "verbose", optional_argument,	NULL, 'v'},
			{ "help",    no_argument,	NULL, 'h'},
			{ "version", no_argument,	NULL, 'V'},
			{ "copying", no_argument,	NULL, 'C'},
			{ "?",	     no_argument,	NULL, 'h'},
			{ 0, }
		};
		/* *INDENT-ON* */

		c = getopt_long_only(argc, argv, "b:f:t:qv::hVC?", long_options, &option_index);
#else				/* _GNU_SOURCE */
		c = getopt(argc, argv, "b:f:t:qv::hVC?");
#endif				/* _GNU_SOURCE */
		if (c == -1)
			break;
		switch (c) {
		case 'b':	/* -b, --burst BURST */
			initial_burst = strtol(optarg, NULL, 0);
			break;
		case 'f':	/* -f, --factor FACTOR */
			time_factor = strtol(optarg, NULL, 0);
			break;
		case 't':	/* -f, --threads THREADS */
			max_threads = strtol(optarg, NULL, 0);
			if (max_threads < 0)
				max_threads = 2;
			else if (max_threads & 0x01)
				max_threads++;
			if (max_threads > MAX_THR)
				max_threads = MAX_THR;
			break;
		case 'q':	/* -q, --quiet */
			verbose = 0;
			break;
		case 'v':	/* -v, --verbose [LEVEL] */
			if (!optarg)
				verbose++;
			else
				verbose = strtol(optarg, NULL, 0);
			break;
		case 'h':	/* -h, --help, -?, --? */
			help(argc, argv);
			exit(0);
		case 'V':	/* -V, --version */
			version(argc, argv);
			exit(0);
		case 'C':	/* -C, --copying */
			copying(argc, argv);
			exit(0);
		case '?':
		default:
			optind--;
		      bad_nonopt:
			if (optind < argc && verbose) {
				fprintf(stderr, "%s: illegal syntax -- ", argv[0]);
				for (; optind < argc; optind++)
					fprintf(stderr, "%s ", argv[optind]);
				fprintf(stderr, "\n");
			}
			usage(argc, argv);
			exit(2);
		}
	}
	if (optind < argc)
		goto bad_nonopt;
}

/************************************************************************
*                            main                                       *
*************************************************************************
*									*
* The main program sets up all the structurs and file descriptors and	*
* then lets the threads just move the data.				*
*									*
************************************************************************/
int
main(int argc, char **argv)
{
	int rslt;
	int i;
	void *(*funcp) (void *);
	thread_info_t *tp;

	get_options(argc, argv);

	for (i = 1; i <= max_threads; i++) {
		if (i & 0x01)
			funcp = reader;
		else
			funcp = writer;

		setup_thread_info(i);
		tp = &thread_info[i];
		rslt = pthread_create(&thread_ids[i], NULL, funcp, tp);
		if (rslt < 0) {
			if (verbose)
				fprintf(stderr, "Thread #%d: ", i);
			if (verbose)
				perror("pthread_create");
		} else if (verbose > 1)
			printf("main: Thread #%d id=%ld\n", i, thread_ids[i]);
	}

	for (i = 1; i <= max_threads; i++) {
		tp = &thread_info[i];
		if (!tp->looped) {
			if (set_loop(tp->fd, tp->other->minor) < 0)
				exit(1);
			tp->looped = 1;
			tp->other->looped = 1;
		}
		if (set_timer(tp->fd, tp->thread_number) < 0)
			exit(1);
	}

	start_time = time(NULL);
	for (i = 1; i <= max_threads; i++) {
		tp = &thread_info[i];
		pthread_mutex_unlock(&tp->startup);
	}

	for (;;) {
		if (verbose > 1)
			printf("main: sleeping for 5 secs\n");
		sleep(5);
		print_progress();
	}

	if (verbose)
		printf("main: exiting\n");

	return (0);
}
